### ISSUE TITLE
Use gold as the linker when available

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#1476]
http://blogs.s-osg.org/an-introduction-to-accelerating-your-build-with-clang/ shows that one can get nice speedup by using gold instead ld as the linker. On manycore machines (e.g., my 64-core build servers) it usually takes about 5 minutes to compile everything, plus 2 minutes to link everything. The last phase of course is sequential, so I *really* case about link time.

https://bugs.webkit.org/show_bug.cgi?id=137953 seems to have some useful ideas.

**Comment by bangerth** [comment]
On my laptop, with a fast SSD:

Regular linking: 25.9 seconds
With -fuse-ld=gold: 10.3 seconds.

I'm very impressed.

**Comment by tamiko** [comment]
As simple as:
```
ENABLE_IF_SUPPORTED(DEAL_II_LINKER_FLAGS "-fuse-ld=gold")
```

**Comment by tamiko** [comment]
Id' suggest we enable it for recent gcc compiler version (in the hope that binutils that ships ld.gold is also recen one), and see whether we encounter some problems.

**Comment by kronbichler** [comment]
I've been using the gold linker for several years on my machines (by putting a symlink to /usr/bin/ld.gold on a path before the system ld; that way, it circumvents possible problems with gold when compiling system libraries from root that have been reported) and can confirm the speedup. If the compiled files are in RAM (cached), the link times are more than 3x better for linking the library. Even better, the test suite runs in less than half the time on my machine :-)

**Comment by bangerth** [comment]
Just for more comparison data:

Regular build+link with ld.bfd:
```
ninja clean
time ninja

real    5m14.030s
user    89m49.428s
sys     10m32.232s
```
With ld.gold:
```
real    3m11.617s
user    89m51.832s
sys     15m41.924s
```
This is about 2 minutes faster.

Next, just relinking the two shared libs with ld.bfd:
```
touch source/grid/CMakeFiles/obj_grid.*.dir/tria.cc.o*
time ninja

real    2m51.833s
user    0m40.092s
sys     2m41.424s
```
And with ld.gold:
```
real    0m18.296s
user    0m10.892s
sys     0m2.056s
```
This by itself is pretty amazing. It's also one of the most often done operations: compile a single file, then build the entire library. This takes me ~3 minutes so far and could be done in around 30 seconds!

Finally, testing and then running the testsuite again without having to compile all test files again with ld.bfd:
```
ninja setup_tests
time ctest -j60

real    27m9.635s
user    871m35.524s
sys     232m47.588s


touch lib/lib*
time ctest -j60

real    15m0.726s
user    418m34.556s
sys     211m20.780s
```


gold:
.....
And this is with ld.gold:
```
real    22m18.594s
user    680m10.772s
sys     236m20.568s

real    9m41.981s
user    224m10.572s
sys     177m17.900s
```
All in all, this is quite impressive!

(For completeness, these numbers were not done using `-fuse-ld=gold` since the compiler on this 64-bit machine doesn't have a GCC that allows this. Instead, I put `-B/path/to/ld` into `CXXFLAGS`, where `/path/to/ld` is a path to an executable `ld` that links to `gold`.)

**Comment by bangerth** [comment]
Fixed by #1539 (which superseded #1492).
